### ISSUE TITLE
Python module does not properly handle MFA/OTP authentication

### ISSUE THREAD

**Issue by webbrandon** [OpenVPN/openvpn3-linux#28]
**Q**: How can I autoload user and password with OTP based MFA dynamic challenge for `openvpn3` and only prompt for MFA code? 

Attempts Made:
----------

I created the below script to inject OTP based MFA code at request time into a `.autoload` config with `openvpn3-autoload` to simplify the login process with our OpenVPN requirements. I figure the script should be able to get around the static requirement until the MFA needs renewal but I am still getting issues with the `dynamic_challenge` parameter. 

I get the following message from connection with the `openvpn3-autoload` using `openvpn3 sessions-list`:

> Status: Connection, Configuration requires user input: Dynamic Challenge

**OpenVPN3 Autoload Credentials Script**

```
    #!/bin/bash
    set -e
    
    ## Requirements
    # - openvpn3
    # - openvpn3-autoload
    
    # When you are logged in you may download a client.ovpn file.  First create a
    # subdir at ~/.openvpn and copy the downloaded file to this directory.  
    
    echo '{"autostart": true,"user-auth": { "autologin": true,"username": "{{ OPENVPN_USR }}", "password": "{{ OPENVPN_PW }}", "dynamic_challenge": "{{ OPENVPN_MFA_CODE }}"},"tunnel": {"ipv6": [ "default" ],"persist": true}}' > ${HOME}/.openvpn/autoload.tmpl
    
    OPENVPN_CLIENT=client
    OPENVPN_USR="my.user"
    OPENVPN_PW="password"
    OPENVPN_USR=${OPENVPN_USR:-`read -p "User login: " OPENVPN_USR; echo $OPENVPN_USR`}
    OPENVPN_PW=${OPENVPN_PW:-`read -p "User Pass: " OPENVPN_PW; echo $OPENVPN_PW`}
    
    read -p "Please enter MFA code: " OPENVPN_MFA_CODE
    
    export OPENVPN_USR=${OPENVPN_USR}
    export OPENVPN_PW=${OPENVPN_PW}
    export OPENVPN_MFA_CODE=${OPENVPN_MFA_CODE}
    
    ## This is an internal tool to our org but you could figure out how to use jq also.
    ## tmpz -t ${HOME}/.openvpn/autoload.tmpl -o ${HOME}/.openvpn/${OPENVPN_CLIENT}.autoload
    rm ${HOME}/.openvpn/autoload.tmpl
    
    # I am open to suggestions if you have a better way.
    openvpn3-autoload --directory ${HOME}/.openvpn
```

I have tried passing a parameter file like below to the *ovpn* file with `auth-user-pass credentials.txt`(ie: dbus).

**credentials.txt**

```
    my.user
    password
```

I get connection failure with `auth-user-pass` and no dynamic challenge.

**Comment by dsommers** [comment]
To be honest, I'm puzzled by your approach ... I would rather recommend you modify or write your own `openvpn3-autoload` variant solving your issue more directly than wrapping a [Python script](https://github.com/OpenVPN/openvpn3-linux/blob/master/src/python/openvpn3-autoload) from a bash script.

I would suggest adding your MFA injection somewhere close to this line: https://github.com/OpenVPN/openvpn3-linux/blob/1888b9711ecd988870e4c17fdeb983c2f82c8b10/src/python/openvpn3-autoload#L275

You can also have a look at the [src/tests/python/start-tunnel2.py](https://github.com/OpenVPN/openvpn3-linux/blob/master/src/tests/python/start-tunnel2.py) test code which covers how to start a pre-imported (via `openvpn3 config-import`) configuration profile.  Configuration file import can also be done in Python, see the [src/tests/python/configimport.py](https://github.com/OpenVPN/openvpn3-linux/blob/master/src/tests/python/configimport.py) for details.  Adding a configuration/option parser (which is used by the `openvpn3-autoload` and `openvpn2` scripts) is demoed here in [src/tests/python/genconfig.py](https://github.com/OpenVPN/openvpn3-linux/blob/master/src/tests/python/genconfig.py).

**Comment by webbrandon** [comment]
This can be done many ways it seems.  I am looking to share this and be configured with minimal effort so python script may be more helpful in achieving that.  

I created python script to do this using the [src/tests/python/genconfig.py](https://github.com/OpenVPN/openvpn3-linux/blob/master/src/tests/python/genconfig.py) example, I still am seeing errors when I call its equivalent `--static-challenge`.  I did notice it embedded my `auth-user-pass` as XML like syntax in the config. I guess with a little bit of unwanted effort I can make some modifications with a python script to process just the MFA question, just need to figure that out in code I guess.

> session-start: ** ERROR ** Failed to start new session: Failed calling D-Bus method Ready: GDBus.Error:net.openvpn.v3.sessions.error: Backend VPN process have died.  Session is no longer valid.

**Comment by dsommers** [comment]
I've been doing some more investigation, and it looks like there is a regression in the Python module which does not properly handle MFA authentication

**Comment by dsommers** [comment]
This should work now with the `openvpn2` implementation in the v16_beta release (with commit 2aa087cbc22497d70d90fca72c06979181a904a3).

The `openvpn3-autoload` approach is questionable in regards to 2FA, and this utility is going to be deprecated in future releases as we have added a better way for handling this with [`systemctl {start,stop,status} openvpn3-session@CONFIGNAME.service`](https://github.com/OpenVPN/openvpn3-linux/blob/master/docs/man/openvpn3-systemd.8.rst#example).

Static and dynamic challenges in `openvpn3-autoload` will not have a high priority going forward.

Closing this issue as fixed.
